vizro-core/src/vizro/models/_controls/filter.py

```diff
@@ -108,31 +110,57 @@ def build(self):
         return self.selector.build()
 
     def _set_targets(self):
-        if not self.targets:
-            for component_id in model_manager._get_page_model_ids_with_figure(
-                page_id=model_manager._get_model_page_id(model_id=ModelID(str(self.id)))
-            ):
-                # TODO: consider making a helper method in data_manager or elsewhere to reduce this operation being
-                #  duplicated across Filter so much, and/or consider storing the result to avoid repeating it.
-                #  Need to think about this in connection with how to update filters on the fly and duplicated calls
-                #  issue outlined in https://github.com/mckinsey/vizro/pull/398#discussion_r1559120849.
-                data_source_name = model_manager[component_id]["data_frame"]
-                data_frame = data_manager[data_source_name].load()
-                if self.column in data_frame.columns:
-                    self.targets.append(component_id)
-            if not self.targets:
-                raise ValueError(f"Selected column {self.column} not found in any dataframe on this page.")
+        # TODO: consider moving to data_manager, depending on Petar's work
+        # TODO: write tests
+        potential_targets = self.targets or model_manager._get_page_model_ids_with_figure(
+            page_id=model_manager._get_model_page_id(model_id=ModelID(str(self.id)))
+        )
+
+        potential_target_to_data_source_name = {
+            target: model_manager[target]["data_frame"] for target in potential_targets
+        }
+        # Using set() here ensures we only load each data source once rather than repeating the operation for each
+        # target.
+        data_source_name_to_data = {
+            data_source_name: data_manager[data_source_name].load()
+            for data_source_name in set(potential_target_to_data_source_name.values())
+        }
+        target_to_series = dict()
+
+        for target, data_source_name in potential_target_to_data_source_name.items():
+            data_frame = data_source_name_to_data[data_source_name]
+
+            if self.column in data_frame.columns:
+                target_to_series[target] = data_frame
+            elif self.targets:
+                # targets were manually specified so it's not ok the column isn't there. If targets were not specified
+                # then it's fine, we just skip this target, and error is not raised.
+                raise ValueError(f"Selected column {self.column} not found in dataframe for {target}.")
+
+        if not target_to_series:
+            raise ValueError(f"Selected column {self.column} not found in any dataframe on this page.")
+
+        self.targets = list(target_to_series)
+        # COMMEnt. will have repeats
+        self._targeted_data = pd.DataFrame.from_dict(target_to_series)
```

Comment:
I don't understand this line, you create a DF that itself has DFs inside it?

Reply:
I assume this somewhat goes like: check all columns, append them all to a big series, then do checks for dtype and alter min max etc?

Reply:
I saw later that this was clearer in a later commit